Merge latest improvements into production

tcouch opened the pull request:
## Changes
### Include unlinked fragments in @mentions  
* Issue: #295 
* PR: #305
Unlinked fragments can now be added via @ mentions. This works in a similar fashion to anonymous fragments search in that a keyword consisting of the letter "u" will find all unlinked fragments, and "u<int>" will find all unlinked fragments whose id starts with that number; e.g. "u33" will find Unlinked 338 and Unlinked 339.

### Work fragments ordered by book
* Issue: #160 
* PR: #304 
Material listed on a work's detail page is now organised by book. Anchor links have been added from the books listed at the top to the area of the page containing material for those books.
#### Examples:
* https://www.ucl.ac.uk/history/frrant-preprod/work/37/
* https://www.ucl.ac.uk/history/frrant-preprod/work/27/

### Fix display of apposita for unlinked fragments
* Issue: #289 
* PR: #303 
#### Examples:
* https://www.ucl.ac.uk/history/frrant-preprod/anonymous/53/

### Ignore angle brackets in search queries
* PR: #302
#### Examples:
* https://www.ucl.ac.uk/history/frrant-preprod/search/?q=findthem&what=all 

### Refuse connections to unexpected locations
* PR: #307 
Technical fix to the webserver refuses connection to unrecognised locations

tcouch committed:
Add context data with material organised by book and link certainty

tcouch committed:
Use book str as key in ordered materials

tcouch committed:
return context not ordered_materials

tcouch committed:
Rearrange ordered_materials dict heirarchy with def/poss under type

tcouch committed:
add query func to strip <> character codes

tcouch committed:
Added global flag to regexp_replace

tcouch committed:
Only show book heading if it has material

tcouch committed:
Remove a print statement;

tcouch committed:
Add anchor links to book sections from work detail

tcouch committed:
Needed to add get_display_string method

tcouch committed:
Get anchor links working properly

tcouch committed:
refactor: simplify and use book objects as keys

tcouch committed:
Remove old content and add padding

tim-band committed:
Merge pull request #302 from UCL/fix/ignore_angle_brackets_in_search

add query func to strip <> character codes

tcouch committed:
If unlinked, show apposita without going through link

tcouch committed:
Show unlinked fragments in appositum links

tcouch committed:
Update docstring

tcouch committed:
Find unlinked with "u" and id

tcouch committed:
Test unlinked fragment mentions

tcouch committed:
Merge pull request #303 from UCL/fix/display_unlinked_apposita

Fix/display unlinked apposita

tcouch committed:
Merge pull request #305 from UCL/feature/include_unlinked_in_mentions

Feature/include unlinked in mentions

tcouch committed:
order materials wrt antiquarian too

tcouch committed:
Merge pull request #304 from UCL/work_fragments_ordered_by_book

Work fragments ordered by book

tcouch committed:
Migration required after changing verbose name

tcouch committed:
Distinguish development server emails from prod

tcouch committed:
Fix line length too long

tcouch committed:
Fix trailing whitespace

Commit:
reject requests we don't expect

tcouch committed:
Merge pull request #306 from UCL/distinctive-dev-emails

Distinguish development server emails from prod

tcouch committed:
Reject unknown locs in prod too

tcouch committed:
Merge pull request #307 from UCL/fix/broken-link-emails

Fix/broken link emails

tcouch commented:
Hi @rmamarshall. The changes listed above are now ready to be added to production. Could you review these on the [development server](https://www.ucl.ac.uk/history/frrant-preprod/) and let me know if you're happy for us to add them please?

rmamarshall commented:
Hi @tcouch. The five changes look really good and everything works as expected, except "Fix display of apposita for unlinked fragments". I turned a fragment with apposita into an unlinked fragment, and the apposita were retained. However, when I next turned the new unlinked fragment into an anonymous fragment, the links disappeared. As Anonymous fragments can also have apposita, this may be an issue. A similar issue was encountered when I turned an unlinked fragment into a non-anonymous fragment: all apposita links disappeared. I think we may have discussed this issue before - apologies if I have got the wrong end of the stick and the above fix is for a different feature!

tcouch commented:
Hi @rmamarshall. Thanks for the feedback. Can I just clarify a couple of things?

I was under the impression that Anonymous Fragments can be apposita to Fragments, but they can't be apposita to each other. I presumed this was a deliberate decision as that's how the database has been designed. So, when converting an unlinked fragment with apposita into an anonymous fragment, it would have to lose its links to other anonymous fragments.

When you say "when I turned an unlinked fragment into a non-anonymous fragment...", did you mean something else? I was confused because, unlinked fragments **are** non-anonymous fragments.

rmamarshall commented:
Hi @tcouch. Apologies, I did indeed have the wrong end of the stick: Anonymous fragments should not have apposita. The other issue I described arose from turning an unlinked fragment into an anonymous one, then turning the anonymous fragment into a non-anonymous one: because of the intermediate step, the apposita were lost, but I don't see why this should happen in actual usage (unliked fragments will simply be linked to authors / works directly). Did we discuss a possible warning pop-up at some point, alerting users to the fact that making a fragment with apposita into an anonymous fragment will remove the apposita links?

tcouch committed:
Warn apposita links will be lost

tcouch committed:
Merge pull request #312 from UCL/feature/unlinked_to_anon_warning

Warn apposita links will be lost when converting unlinked fragments

raquelalegre commented on src/config/settings/development.py:
```diff
@@ -97,13 +97,14 @@ def monkey_patch():
 DEFAULT_FROM_EMAIL = env(
     "DJANGO_DEFAULT_FROM_EMAIL",
-    default="Republican Antiquarians Research Database <noreply@example.com>",
+    default="Republican Antiquarians Research Database (dev) <noreply@example.com>",
```
It shouldn't say "(dev)" in the production version, right? This line shouldn't be merged.

tcouch replied:
This settings file is only used by the development server. The production server uses src/config/settings/production.py which has these variables set as suggested.

raquelalegre replied:
Ah yes :) Silly me!

raquelalegre commented on src/config/settings/development.py:
```diff
@@ -97,13 +97,14 @@ def monkey_patch():
 EMAIL_SUBJECT_PREFIX = env(
-    "DJANGO_EMAIL_SUBJECT_PREFIX", default="[Republican Antiquarians Research Database]"
+    "DJANGO_EMAIL_SUBJECT_PREFIX",
+    default="[Republican Antiquarians Research Database (dev)]",
```
Same as earlier comment.

tcouch replied:
See my reply to your comment above.

raquelalegre reviewed (approved):
Looks great. Just made 2 small comments.

tcouch commented:
Hi @rmamarshall I've updated the warning message to mention that apposita links will be lost and added that to this pull request. Are you happy for me to go ahead and put this into production now? If so you can click to approve these changes, or just reply to this message.

rmamarshall commented:
Hi @tcouch Happy New Year! I'm not sure I understand - is this ready to trial in the pre-production server, or will it simply be added to the production version?

tcouch committed:
Update link to quill-mention css and js files

tcouch committed:
Merge pull request #313 from UCL/fix/missing-quill-mention

Update link to quill-mention css and js files

tcouch commented:
Hi @rmamarshall, yes you can test this out on the pre-production server now. If you're happy with it, let me know and I'll put it into production.

tcouch commented:
Hi @rmamarshall, can you confirm for me if you're happy for these changes to be merged into production now?

rmamarshall commented:
Hi @tcouch I've just checked the new feature in the pre-production server and it seems to work perfectly; please merge it into production.
